lib/core/test/unit/Cardano/WalletSpec.hs

```diff
@@ -202,10 +211,93 @@ spec = do
+            W.guardJoin knownPools dlg pidA `shouldBe` Right ()
+        it "Cannot join A, when active = A, next = [B, A]" $ do
+            let next1 = next (EpochNo 1) (Delegating pidB)
+            let next2 = next (EpochNo 2) (Delegating pidA)
+            let dlg = WalletDelegation
+                    {active = Delegating pidA, next = [next1, next2]}
+            W.guardJoin knownPools dlg pidA
+                `shouldBe` Left (W.ErrAlreadyDelegating pidA)
+        it "Cannot join when pool is unknown" $ do
+            let dlg = WalletDelegation {active = NotDelegating, next = []}
+            W.guardJoin knownPools dlg pidUnknown
+                `shouldBe` Left (W.ErrNoSuchPool pidUnknown)
+        it "Cannot quit when active: not_delegating, next = []" $ do
+            let dlg = WalletDelegation {active = NotDelegating, next = []}
+            W.guardQuit dlg `shouldBe` Left (W.ErrNotDelegatingOrAboutTo)
+        it "Cannot quit when active: A, next = [not_delegating]" $ do
+            let next1 = next (EpochNo 1) NotDelegating
+            let dlg = WalletDelegation {active = Delegating pidA, next = [next1]}
+            W.guardQuit dlg `shouldBe` Left (W.ErrNotDelegatingOrAboutTo)
+        it "Cannot quit when active: A, next = [B, not_delegating]" $ do
+            let next1 = next (EpochNo 1) (Delegating pidB)
+            let next2 = next (EpochNo 2) NotDelegating
+            let dlg = WalletDelegation
+                    {active = Delegating pidA, next = [next1, next2]}
+            W.guardQuit dlg `shouldBe` Left (W.ErrNotDelegatingOrAboutTo)
+        it "Can quit when active: not_delegating, next = [A]" $ do
+            let next1 = next (EpochNo 1) (Delegating pidA)
+            let dlg = WalletDelegation
+                    {active = NotDelegating, next = [next1]}
+            W.guardQuit dlg `shouldBe` Right ()
+     where
+         pidA = PoolId "A"
+         pidB = PoolId "B"
+         pidUnknown = PoolId "unknown"
+         knownPools = [pidA, pidB]
+         next epoch dlgStatus =
+             WalletDelegationNext {changesAt = epoch, status = dlgStatus}
+
+
 {-------------------------------------------------------------------------------
                                     Properties
 -------------------------------------------------------------------------------}
 
+prop_guardJoinQuit
+    :: [PoolId]
+    -> WalletDelegation
+    -> PoolId
+    -> Property
+prop_guardJoinQuit knownPools dlg pid =
+    case W.guardJoin knownPools dlg pid of
+        Right () -> label "I can join" $ property True
+        Left W.ErrNoSuchPool{}  -> label "ErrNoSuchPool" $ property True
+        Left W.ErrAlreadyDelegating{} ->
+            label "ErrAlreadyDelegating" (W.guardQuit dlg === Right ())
+
+prop_guardQuitJoin
+    :: [PoolId]
+    -> WalletDelegation
+    -> PoolId
+    -> Property
+prop_guardQuitJoin knownPools dlg pid =
+    pid `elem` knownPools ==> case W.guardQuit dlg of
```

Comment:
It can be _"dangerous"_ to use a pre-condition for filtering generated data. It can lead to false-negative where QC will give-up if it has to discard too many cases (and will assume that the pre-condition can't be met). 

A better practice would be to generate the right data in the first place. Here, the property doesn't really have any assertions the `pid` or the list of pools. So, instead of generating them separately, one could generate a `NonEmpty PoolId` and pick a pool id from it (and it's guaranteed to be in the list :grimacing:)!